HW_1/hw_1.py

```diff
@@ -0,0 +1,466 @@
+
+Вся сложность задания в том, чтобы правильно организовать код. Если у вас есть повторяющийся код в сестринских классах или родительском и дочернем, значит вы что-то делаете не так.
+
+
+Маленькое замечание: По-хорошему, между классом `BiologicalSequence` и классами `NucleicAcidSequence` и `AminoAcidSequence`, ещё должен быть класс-прослойка, частично реализующий интерфейс `BiologicalSequence`, но его писать не обязательно, так как задание и так довольно большое (правда из-за этого у вас неминуемо возникнет повторяющийся код в классах `NucleicAcidSequence` и `AminoAcidSequence`)
+"""
+
+from abc import ABC, abstractmethod
+
+class BiologicalSequence(ABC):    
+    @abstractmethod
+    def __len__(self):
+        pass
+    
+    @abstractmethod
+    def __getitem__(self, ind):
+        pass
+
+    @abstractmethod
+    def __repr__(self):
+        pass
+    
+    @abstractmethod
+    def __str__(self):
+        pass
+    
+    @abstractmethod
+    def check_alphabet(self):
+        pass
+
+class NucleicAcidSequence(BiologicalSequence):
+
+    dna_complement = {"A": "T", "a": "t",
+                      "T": "A", "t": "a",
+                      "G": "C", "g": "c",
+                      "C": "G", "c": "g"}
+
+    rna_complement = {"A": "U", "a": "u",
+                      "U": "A", "u": "a",
+                      "G": "C", "g": "c",
+                      "C": "G", "c": "g"}
+
+    def __init__(self, seq:str):
+        self.seq = seq
+    
+    def __len__(self):
+        return len(self.seq)
+
+    def __getitem__(self, ind):
+        return self.seq.__getitem__(ind)
+    
+    def __repr__(self):
+        return repr(self.seq)
+    
+    def __str__(self):
+        return str(self.seq)
+    
+    def check_alphabet(self, dna_complement = dna_complement, rna_complement = rna_complement):
+        acid = 0
+        for i in self.seq:
+            if i in dna_complement and i not in rna_complement:
+                continue
+            elif i not in dna_complement and i in rna_complement:
+                acid = 1
+            elif i in dna_complement and i in rna_complement:
+                continue
+            else:
+                raise SyntaxError("Wrong alphabet. Please rewrite sequence")
+        if acid != 0:
+            return "RNA"
+        else:
+            return "DNA" 
+    
+    
+    def complement(self, dna_complement = dna_complement, rna_complement = rna_complement):
+        compl_seq = []
+        if self.check_alphabet() == "DNA":
+            for i in self.seq:
+                compl_seq.append(dna_complement[i])
+        elif self.check_alphabet() == "RNA":
+            for i in self.seq:
+                compl_seq.append(rna_complement[i])
```

Comment:
Гипотетическая ситуация. У нас есть 50 разных видов нуклеиновых кислот. Писать 50 if-elif это совсем не круто. Гораздо удобнее для каждого из комплементируемых объектов сделать атрибут с таблицей для комплементирования (например, `complement_table`) и использовать его здесь. В таком случае получится что-то вот такое, это гораздо гибче

```suggestion
    for nucl in self.seq:
        compl_seq.append(self.complement_table[nucl])
```